# features explaination

Health & Mortality Indicators

*   Adult_mortality: deaths of adults per 1000 population.
*   Infant_deaths: Number of infant deaths per 1000 live births.
*   Under_five_deaths: Deaths of children under 5 per 1000 population.
*   Hepatitis_B: Percentage of immunization coverage against Hepatitis B.
*   Polio: Percentage of immunization coverage against Polio.
*   Diphtheria: Percentage of immunization coverage against Diphtheria among 1-year-olds.
*   Measles: Percentage of immunization coverage against Measles among 1-year-olds.
*   Incidents_HIV: Deaths caused by HIV/AIDS per 1000 population.

Lifestyle & Physical Health

*   Alcohol_consumption: The average annual alcohol consumption in liters per individual +15 years old.
*   BMI: Average Body Mass Index.
*   Thinness_ten_nineteen_years: Malnutrition indicator for youth (10–19 years).
*   Thinness_five_nine_years: Malnutrition indicator for children (5–9 years).

Economic & Socioeconomic Indicators

*   GDP_per_capita: The average income generated per person in a country.
*   Population_mln: The total population size measured in millions.
*   Schooling: Average years of schooling.

Economic Status

*   Economy_status_Developed: Indicates whether a country is classified as developed.
*   Economy_status_Developing: Indicates whether a country is classified as developing.

Target Variable

*   Life_expectancy: Average number of years a person is expected to live at birth.


#Import Packages

In [40]:
import numpy as np
import pandas as pd #To read csv file
import matplotlib.pyplot as plt
#import seaborn as sns #To plot Histograms with Distribution curves
from matplotlib import axes #Organize feature sub-groups
#unsure
import scipy.stats as st

In [41]:
data = pd.read_csv('Dataset/Life-Expectancy-Data-Updated.csv')

In [42]:
data

Country          Region  Year  Infant_deaths  Under_five_deaths  \
0       Turkiye     Middle East  2015           11.1               13.0   
1         Spain  European Union  2015            2.7                3.3   
2         India            Asia  2007           51.5               67.9   
3        Guyana   South America  2006           32.8               40.5   
4        Israel     Middle East  2012            3.4                4.3   
...         ...             ...   ...            ...                ...   
2859      Niger          Africa  2000           97.0              224.9   
2860   Mongolia            Asia  2009           23.9               28.6   
2861  Sri Lanka            Asia  2004           17.7               28.9   
2862  Lithuania  European Union  2002            7.9                9.9   
2863    Iceland  Rest of Europe  2011            2.1                2.6   

      Adult_mortality  Alcohol_consumption  Hepatitis_B  Measles   BMI  ...  \
0            105.8240                1.320           97       65  27.8  ...   
1             57.9025               10.350           97       94  26.0  ...   
2            201.0765                1.570           60       35  21.2  ...   
3            222.1965                5.680           93       74  25.3  ...   
4             57.9510                2.890           97       89  27.0  ...   
...               ...                  ...          ...      ...   ...  ...   
2859         291.8240                0.092           72       64  20.8  ...   
2860         235.2330                6.560           97       97  25.3  ...   
2861         134.8950                1.560           62       95  21.9  ...   
2862         204.0120               11.000           94       95  26.1  ...   
2863          50.5745                6.840           88       90  26.1  ...   

      Diphtheria  Incidents_HIV  GDP_per_capita  Population_mln  \
0             97           0.08           11006           78.53   
1             97           0.09           25742           46.44   
2             64           0.13            1076         1183.21   
3             93           0.79            4146            0.75   
4             94           0.08           33995            7.91   
...          ...            ...             ...             ...   
2859          34           0.49             399           11.33   
2860          95           0.02            2515            2.67   
2861          97           0.02            2130           19.39   
2862          95           0.05            7424            3.44   
2863          95           0.05           48934            0.32   

      Thinness_ten_nineteen_years  Thinness_five_nine_years  Schooling  \
0                             4.9                       4.8        7.8   
1                             0.6                       0.5        9.7   
2                            27.1                      28.0        5.0   
3                             5.7                       5.5        7.9   
4                             1.2                       1.1       12.8   
...                           ...                       ...        ...   
2859                         12.8                      12.9        1.1   
2860                          2.2                       2.3        9.1   
2861                         15.4                      15.5       10.3   
2862                          3.3                       3.3       11.1   
2863                          0.9                       0.9       11.0   

      Economy_status_Developed  Economy_status_Developing  Life_expectancy  
0                            0                          1             76.5  
1                            1                          0             82.8  
2                            0                          1             65.4  
3                            0                          1             67.0  
4                            1                          0             81.7  
..

Check the dimension of the dataset

In [43]:
data.shape

(2864, 21)

2864 rows → number of observations (n)

21 columns → number of variables (features + target)



Get information about the data

In [44]:
# Display column names and data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2864 entries, 0 to 2863
Data columns (total 21 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Country                      2864 non-null   object 
 1   Region                       2864 non-null   object 
 2   Year                         2864 non-null   int64  
 3   Infant_deaths                2864 non-null   float64
 4   Under_five_deaths            2864 non-null   float64
 5   Adult_mortality              2864 non-null   float64
 6   Alcohol_consumption          2864 non-null   float64
 7   Hepatitis_B                  2864 non-null   int64  
 8   Measles                      2864 non-null   int64  
 9   BMI                          2864 non-null   float64
 10  Polio                        2864 non-null   int64  
 11  Diphtheria                   2864 non-null   int64  
 12  Incidents_HIV                2864 non-null   float64
 13  GDP_per_capita    

Classification of Variables in the Dataset

In [45]:
# Categorical variables
categorical_variables = ["Country", "Region"]

# Numerical variables
numerical_variables = data.columns.drop(categorical_variables)

numerical_variables

Index(['Year', 'Infant_deaths', 'Under_five_deaths', 'Adult_mortality',
       'Alcohol_consumption', 'Hepatitis_B', 'Measles', 'BMI', 'Polio',
       'Diphtheria', 'Incidents_HIV', 'GDP_per_capita', 'Population_mln',
       'Thinness_ten_nineteen_years', 'Thinness_five_nine_years', 'Schooling',
       'Economy_status_Developed', 'Economy_status_Developing',
       'Life_expectancy'],
      dtype='object')

In [46]:
data["Country"].unique()

array(['Turkiye', 'Spain', 'India', 'Guyana', 'Israel', 'Costa Rica',
       'Russian Federation', 'Hungary', 'Jordan', 'Moldova', 'Brazil',
       'Malta', 'Bahamas, The', 'Ukraine', 'Switzerland', 'Norway',
       'Finland', 'Comoros', 'Japan', 'Gabon', 'Ghana', 'Philippines',
       'Congo, Rep.', 'Madagascar', 'Estonia', 'Belize', 'Kazakhstan',
       'Cameroon', 'Zimbabwe', 'Bhutan', 'South Africa', 'Eritrea',
       'Germany', 'Saudi Arabia', 'Kiribati', 'Seychelles', 'Singapore',
       'Togo', 'Denmark', 'Gambia, The', 'Sweden', 'Austria',
       'Kyrgyz Republic', 'Grenada', 'Brunei Darussalam', 'Greece',
       'Uruguay', 'Croatia', 'Romania', 'Central African Republic',
       'Algeria', 'Yemen, Rep.', 'Armenia',
       'St. Vincent and the Grenadines', 'Kenya', 'Micronesia, Fed. Sts.',
       'Antigua and Barbuda', 'Nepal', 'Lithuania', 'Vanuatu',
       'Afghanistan', 'Kuwait', 'Argentina', 'Panama', 'Oman', 'France',
       'Bosnia and Herzegovina', 'Mauritania', 'Somalia

In [47]:
data["Country"].unique().size

179

Country is a nominal categorical variable → have 179 unique categories, would create hundreds of dummy variables so we will drop this column instead of making one hot encoding and the country does not affect the life expectancy

# Drop Country and economy status developing columns

In [48]:
# Drop Country column, also drop one of the economy status to prevent data leakage
data.drop(columns=["Country", "Economy_status_Developing"], inplace=True)
numerical_variables = numerical_variables.drop("Economy_status_Developing")

In [49]:
data["Region"].unique()

array(['Middle East', 'European Union', 'Asia', 'South America',
       'Central America and Caribbean', 'Rest of Europe', 'Africa',
       'Oceania', 'North America'], dtype=object)

In [50]:
data["Region"].unique().size

9

In [ ]:
plt.figure(figsize=(10,6))
sns.boxplot(x='Region', y='Life_expectancy', data=data)
plt.xticks(rotation=45)
plt.title('Life Expectancy Across Regions')
plt.show()


 Region has low cardinality (9 categories), unlike Country.
Keeping this variable allows the model to capture regional differences in life expectancy,
 so one-hot encoding is applied.


# One-Hot Encoding to the ['Region] column


In [51]:
data = pd.get_dummies(
    data,
    columns=['Region'],
    drop_first=True,
    dtype = int
)


In [52]:
data


Year  Infant_deaths  Under_five_deaths  Adult_mortality  \
0     2015           11.1               13.0         105.8240   
1     2015            2.7                3.3          57.9025   
2     2007           51.5               67.9         201.0765   
3     2006           32.8               40.5         222.1965   
4     2012            3.4                4.3          57.9510   
...    ...            ...                ...              ...   
2859  2000           97.0              224.9         291.8240   
2860  2009           23.9               28.6         235.2330   
2861  2004           17.7               28.9         134.8950   
2862  2002            7.9                9.9         204.0120   
2863  2011            2.1                2.6          50.5745   

      Alcohol_consumption  Hepatitis_B  Measles   BMI  Polio  Diphtheria  ...  \
0                   1.320           97       65  27.8     97          97  ...   
1                  10.350           97       94  26.0     97          97  ...   
2                   1.570           60       35  21.2     67          64  ...   
3                   5.680           93       74  25.3     92          93  ...   
4                   2.890           97       89  27.0     94          94  ...   
...                   ...          ...      ...   ...    ...         ...  ...   
2859                0.092           72       64  20.8     41          34  ...   
2860                6.560           97       97  25.3     96          95  ...   
2861                1.560           62       95  21.9     97          97  ...   
2862               11.000           94       95  26.1     97          95  ...   
2863                6.840           88       90  26.1     95          95  ...   

      Economy_status_Developed  Life_expectancy  Region_Asia  \
0                            0             76.5            0   
1                            1             82.8            0   
2                            0             65.4            1   
3                            0             67.0            0   
4                            1             81.7            0   
...                        ...              ...          ...   
2859                         0             49.9            0   
2860                         0             66.9            1   
2861                         0             74.3            1   
2862                         1             71.8            0   
2863                         1             82.4            0   

      Region_Central America and Caribbean  Region_European Union  \
0                                        0                      0   
1                                        0                      1   
2                                        0                      0   
3                                        0                      0   
4                                        0                      0   
...                                    ...                    ...   
2859                                     0                      0   
2860                                     0                      0   
2861                                     0                      0   
2862                                     0                      1   
2863                                     0                      0   

      Region_Middle East  Region_North America  Region_Oceania  \
0                      1                     0               0   
1                      0                     0               0   
2                      0                     0               0   
3                      0                     0               0   
4                      1                     0               0   
...                  ...                   ...             ...   
2859                   0                     0               0   
2860                   0                     0               0   
2861                   0                     0              

# Count missing values for each column


In [53]:
missing_count = data.isnull().sum()
missing_count

Year                                    0
Infant_deaths                           0
Under_five_deaths                       0
Adult_mortality                         0
Alcohol_consumption                     0
Hepatitis_B                             0
Measles                                 0
BMI                                     0
Polio                                   0
Diphtheria                              0
Incidents_HIV                           0
GDP_per_capita                          0
Population_mln                          0
Thinness_ten_nineteen_years             0
Thinness_five_nine_years                0
Schooling                               0
Economy_status_Developed                0
Life_expectancy                         0
Region_Asia                             0
Region_Central America and Caribbean    0
Region_European Union                   0
Region_Middle East                      0
Region_North America                    0
Region_Oceania                    

# Outlier Detection using IQR

In [54]:
outlier_summary = {}  # Dictionary to store the number of outliers for each column

for col in numerical_variables:
    Q1 = data[col].quantile(0.25)  # 25th percentile
    Q3 = data[col].quantile(0.75)  # 75th percentile
    IQR = Q3 - Q1                  # Interquartile range (IQR)

    lower_bound = Q1 - 1.5 * IQR   # Lower bound
    upper_bound = Q3 + 1.5 * IQR   # Upper bound

    # Select rows with value is below lower_bound or above upper_bound
    outliers = data[(data[col] < lower_bound) | (data[col] > upper_bound)]

    outlier_summary[col] = outliers.shape[0]  # Store the count of outliers

# Convert the dictionary to a DataFrame
pd.DataFrame.from_dict(outlier_summary, orient="index", columns=["Outlier Count"])


Outlier Count
Year                                     0
Infant_deaths                           29
Under_five_deaths                      102
Adult_mortality                        112
Alcohol_consumption                      2
Hepatitis_B                            164
Measles                                 35
BMI                                     25
Polio                                  190
Diphtheria                             187
Incidents_HIV                          461
GDP_per_capita                         425
Population_mln                         362
Thinness_ten_nineteen_years             89
Thinness_five_nine_years                95
Schooling                                0
Economy_status_Developed               592
Life_expectancy                         19

In [55]:
# Highly skewed variables (difference between values is important)
cols_to_log = [
    'Infant_deaths',
    'Under_five_deaths',
    'GDP_per_capita',
    'Population_mln',
    'Incidents_HIV'
]

# Health-related continuous variables
cols_to_cap = [
    'BMI',
    'Hepatitis_B',
    'Polio',
    'Diphtheria',
    'Thinness_ten_nineteen_years',
    'Thinness_five_nine_years',
    'Alcohol_consumption',
    'Measles'
]

# Variable requiring row removal
col_to_remove_outliers = [
    'Adult_mortality',
    'Life_expectancy'
]

Apply Log Transformation

In [56]:
for col in cols_to_log:
    data[col] = np.log1p(data[col])  # log1p avoids issues with zero values


Remove Outliers (Adult Mortality, life expectancy)

In [57]:
for col in col_to_remove_outliers:
    # 1. Calculate Bounds for THIS column
    Q1 = data[col].quantile(0.25)
    Q3 = data[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    
    # 2. Filter: Keep rows where THIS column is valid
    data = data[(data[col] >= lower) & (data[col] <= upper)]



Split Data to Target & Features for Regreesion model

In [58]:
# Define target variable
y = data['Life_expectancy']

# Define feature matrix (drop target)
X = data.drop(columns=['Life_expectancy'])


# Split the data into training and testing sets
## 80% training, 20% testing

In [59]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)


Apply IQR-Based Capping

In [60]:
for col in cols_to_cap:
    Q1 = X_train[col].quantile(0.25)
    Q3 = X_train[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    X_train[col] = X_train[col].clip(lower, upper)
    X_test[col] = X_test[col].clip(lower, upper)


Check Remaining Outliers

In [61]:
remaining_outliers = {}

for col in X_train.select_dtypes(include='number').columns:
    Q1 = X_train[col].quantile(0.25)
    Q3 = X_train[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    remaining_outliers[col] = ((X_train[col] < lower) | (X_train[col] > upper)).sum()

pd.DataFrame.from_dict(
    remaining_outliers,
    orient='index',
    columns=['Remaining Outliers']
)


Remaining Outliers
Year                                                   0
Infant_deaths                                          0
Under_five_deaths                                      0
Adult_mortality                                       32
Alcohol_consumption                                    0
Hepatitis_B                                            0
Measles                                                0
BMI                                                    0
Polio                                                  0
Diphtheria                                             0
Incidents_HIV                                        277
GDP_per_capita                                         0
Population_mln                                        28
Thinness_ten_nineteen_years                            0
Thinness_five_nine_years                               0
Schooling                                              0
Economy_status_Developed                             468
Region_Asia                                          358
Region_Central America and Caribbean                 231
Region_European Union                                341
Region_Middle East                                   170
Region_North America                                  36
Region_Oceania                                       144
Region_Rest of Europe                                191
Region_South America                                 154

Descriptive Statistics

In [ ]:
def compute_mean(x):
    return sum(x) / len(x)



In [ ]:
def compute_median(x):
    x_sorted = sorted(x)  # sort values first
    n = len(x_sorted)
    mid = n // 2
    if n % 2 == 0:
        return (x_sorted[mid - 1] + x_sorted[mid]) / 2  # average of two middle values for even data
    else:
        return x_sorted[mid] # middle value for odd data


In [ ]:
def compute_mode(x):
    freq = {}
    for value in x:
        freq[value] = freq.get(value, 0) + 1
    max_freq = max(freq.values())
    modes = [k for k, v in freq.items() if v == max_freq]
    return modes[0]  # return first mode


In [ ]:
def compute_variance(x):
    mean = compute_mean(x)
    return sum((xi - mean)**2 for xi in x) / (len(x) - 1)


In [ ]:
def compute_std(x):
    return compute_variance(x) ** 0.5


In [ ]:
def min_value(x):
    m = x[0]
    for value in x:
        if value < m:
            m = value
    return m

def max_value(x):
    m = x[0]
    for value in x:
        if value > m:
            m = value
    return m


Measures of Central Tendency


In [ ]:
central_tendency = {}

for col in numerical_variables:
    values = data[col].dropna().tolist()
    central_tendency[col] = {
        "Mean": compute_mean(values),
        "Median": compute_median(values),
        "Mode": compute_mode(values)
    }

central_df = pd.DataFrame(central_tendency).T # Convert results to DataFrame to visuallize
central_df


Measures of Dispersion

In [ ]:
dispersion = {}

for col in numerical_variables:
    values = data[col].dropna().tolist()
    dispersion[col] = {
        "Variance": compute_variance(values),
        "Standard Deviation": compute_std(values),
        "Min Value": min_value(values),
        "Max Value": max_value(values)
    }
dispersion_df = pd.DataFrame(dispersion).T
dispersion_df


In [ ]:
data.describe()

Feature Standardization (Z-score)

In [ ]:
# 1. Define the columns to SKIP (The Binary ones)
cols_to_skip = [
    "Economy_status_Developed", 
    'Region_Asia', 'Region_Central America and Caribbean', 'Region_European Union', 
    'Region_Middle East', 'Region_North America', 'Region_Oceania', 
    'Region_Rest of Europe', 'Region_South America',
    'bias' 
]

# 2. The Loop
for col in X_train.columns:
    if col not in cols_to_skip:
        
        # Calculate stats ONLY on Training data
        train_mean = compute_mean(X_train[col])
        train_std = compute_std(X_train[col], train_mean)
        
        if train_std != 0:
            X_train[col] = (X_train[col] - train_mean) / train_std
            X_test[col] = (X_test[col] - train_mean) / train_std
        else:
            X_train[col] = 0
            X_test[col] = 0

In [ ]:
X_train

In [ ]:
# Combine features and target
train_clean = X_train.copy()
train_clean['Life_expectancy'] = y_train

test_clean = X_test.copy()
test_clean['Life_expectancy'] = y_test

# Save to CSV
train_clean.to_csv("train_clean.csv", index=False)
test_clean.to_csv("test_clean.csv", index=False)
